### ISSUE TITLE
Improve keyboard handling

### ISSUE THREAD

**Issue by mantognini** [SFML/SFML#7]
Uniformly manage all keyboards accordingly to their localization.

A (partial) implementation can be found in PR #568, however:

> Its purpose is to be a workaround for the ones in need of scancodes support before the official implementation. And I would be glad if it could help the SFML devs during the creation of the official implementation as well.

**Comment by zsbzsb** [comment]
For the record, I am working on this. :construction_worker:

**Comment by mantognini** [comment]
Shouldn't we first decide the shape of the API? BTW, there was a rather long discussion about that issue somewhere on the forum that unfortunately stalled. Maybe we should restart from there?

**Comment by zsbzsb** [comment]
>there was a rather long discussion about that issue somewhere on the forum that unfortunately stalled. Maybe we should restart from there?

http://en.sfml-dev.org/forums/index.php?topic=13692.msg95809#msg95809

Have at it, I'm working on a practical implementation that will do what it is supposed to do. Once we get something that works we can discuss the API and adapt as needed. The trouble isn't deciding on an API, but getting something that works.

I for one am not up for another long discussion, we already had it and you can read that thread above and all the other threads where stuff has been proposed, the outcome can just about in every case be summed up below.

I have looked at many other libs and how they handle keyboard keys (not talking text input here as what we have works just fine) and they all have several things in common.

- Physical Key (mapped from OS value to the lib value)
- Virtual Key (mapped from OS VK value to lib value) - this is what we have right now
- Modifier Keys (which we already do)

The plan right now is to do the physical key mapping, then fix our current virtual mapping (add missing keys).

**Comment by Bromeon** [comment]
It's nice to have a working implementation as proof-of-concept, but if the API is not thoroughly discussed beforehand, you have to take into account that major changes may be necessary. Same for the mask feature, by the way. We don't want you to waste a lot of effort on something that may in the end not be accepted, as rewriting things again and again can be very frustrating...

Furthermore, I think defining an API as soon as user requirements are known is often a good idea. The API is a user interface and should thus be user-driven, not developer-driven -- the priority is usability, not ease of implementation. Of course there are sometimes implementation details that dictate the API, but it should not be the default approach.

**Comment by binary1248** [comment]
> you have to take into account that major changes may be necessary.

Some people might be willing to take that risk :grin:. If you ask me, it's not for us to decide how they spend their time. If they had fun writing the draft that gets overhauled anyway, there really isn't any loss for anyone.

> We don't want you to waste a lot of effort on something that may in the end not be accepted, as rewriting things again and again can be very frustrating...

Well, _some implementation_ will ultimately get accepted, and if it is even remotely derived from an initial draft or just uses it as inspiration, it will not have been in vain. Getting lost in months of discussions without any substance behind them can be more frustrating than having to write a bit of new code every time a new idea pops up.

> Furthermore, I think defining an API as soon as user requirements are known is often a good idea.

This is assuming we will ever have a good idea about user requirements. I don't like having such dependent development. If you ask me, showing people something that they can try out in a sandbox project will provide us with more information than a simple "what do you think of this discussion?" approach. It obviously hasn't worked for us since this issue is still unresolved even after this amount of discussion.

Everybody should know by now that the waterfall model is busted, and blocking possible parallel work on a draft implementation waiting on such discussions is nothing other than adhering to that model. If we determine that requirements are going to change during the discussions than we should always be able to tweak the draft without too much effort, and this isn't even an issue for us if you ask me.

All we need is "something to talk about" and that something is not an abstract mental model, but something we can send to users and ask them for feedback about, since as you said, the API should be user-driven.

**Comment by Bromeon** [comment]
That's true indeed. As mentioned, a proof-of-concept is definitely helpful. I just thought I'd state such a warning so that people are prepared ;)

One problem with our discussions is -- besides the coexistence of multiple ones about the same topic -- that we start them, discontinue them at some point, and next time we want to work on the feature, everything has been forgotten and needs to be extracted anew from heaps of posts. For larger feature changes, it might thus be good to write a Wiki document that summarizes important considerations and presents different implementation possibilities with their trade-offs.

**Comment by zsbzsb** [comment]
>you have to take into account that major changes may be necessary.

You have to consider we are talking keyboard handling here, not more theory that major changes may have to be done to a general PR that implements some wide open feature. As I said above, all major libs handle it the way I am working to implement something. So unless we all agree on something that makes no sense there won't be any major changes.

>I think defining an API as soon as user requirements are known is often a good idea.

As I also said, I have gone through just about all of the threads requesting better keyboard handling. I have a fairly good idea what people have requested.

>the priority is usability, not ease of implementation.

Yes of course, the current implementation was easy - not much thought given ease of use and multiple keyboard layouts. Trying to be smart and mix virtual key codes and scan codes has given us the current API when it is clear from looking at other implementations it doesn't work out - especially when it comes to handling localizations.

Another thing to think about with this model of don't implement it until we decide is simply won't work. It is impossible to satisfy everybody 100% with every design. There always is going to be drawbacks, it eventually gets to the point where you just need to "suck it up" and go with it. And stalling a merge until we get feedback is fine, but it gets to a point when people could care less and don't bother commenting on anything but in master. So waiting until we get feedback will block forever since feedback won't happen until it is in master and they are basically forced to give usage feedback. Its a sad fact, but that's the way it is.

You all should know me, I love discussing and finding the best design for something and I also don't mind changing something if I found a better way to do it. But at one point you need to end the discussion and just make something work. That is why I have gone ahead and implemented the clipping masks and text outlines.

**Comment by LaurentGomila** [comment]
Personally I like this attitude (and so far, I also like how you do it), this is probably this only way to make these prehistoric issues progress. And I know that you'll be able to deal with whatever we'll do to your PRs, as frustrating as it may be (or not) ;)

:+1:

**Comment by mantognini** [comment]
Those are indeed valid points! Let's see what you'll come up with. :wink:

**Comment by binary1248** [comment]
For posterity reference, ISO/IEC already defined a physical reference grid for keyboards as can be seen [here](https://en.wikipedia.org/wiki/ISO/IEC_9995#Physical_division_and_reference_grid).

**Comment by mantognini** [comment]
Relevant discussion in #1340.

Implementation of scancode: #1235 (missing linux implementation, mainly).

**Comment by eXpl0it3r** [comment]
The keyboard handling has massively been improved by handling scancodes with #1235

As such, I'll close this issue. Further discussions on expanding the non-scancode key codes can be held over at #1045
